dags/tpu-obs/node_pool_util.py

```
@@ -0,0 +1,292 @@
+      if node_name:
+        node_names.append(node_name)
+      else:
+        logging.warning(
+            "Could not extract node name from URL: %s", instance_url
+        )
+  if zone is None:
+    raise RuntimeError(
+        f"No zone found for node pool {info.node_pool_name}."
+    )
+  return node_names, zone
+
+
+@decorators.task
+def delete_node(info: Info):
+  """Defines an Airflow task to delete a random node from the GKE node pool.
+
+  This function uses Airflow's `@task` decorator to create a Python callable
+  that will be executed as an Airflow task. The callable itself performs
+  the node listing, selection, and deletion using `gcloud` commands.
+
+  Args:
+      info (Info): An instance of the Info class containing GKE node pool
+                   configuration parameters.
+
+  Returns:
+      The decorated Airflow task object, ready to be included in a task flow.
+  """
+
+  nodes_list, zone = list_nodes(info)
+  if not nodes_list:
+    logging.warning(
+        "No nodes found in node pool '%s'. No deletion will be performed.",
+        info.node_pool_name,
+    )
+    return None  # Task succeeds, but no node deleted
+
+  node_to_delete = random.choice(nodes_list)
+  logging.info(
+      "Randomly selected node for deletion: %s",
+      node_to_delete,
+  )
+
+  command = f"""
+      gcloud compute instances delete {node_to_delete} \\
+          --project={info.project_id} \\
+          --zone={zone} \\
+          --quiet
+      """
+
+  process = None
+  try:
+    process = subprocess.run(
+      command, shell=True, check=True, capture_output=True, text=True
+  )
+    logging.info("Successfully deleted node %s", node_to_delete)
+  except Exception as e:
+    logger.error('Failed to run "%s": %s', " ".join(command), e)
+    if process:
+      # Subprocess hook stores the stdout + stderr in the `output` attribute
+      logger.debug("STDOUT message: %s", process.stdout)
+      logger.debug("STDERR message: %s", process.stderr)
+    raise
```

consistency, if we use the method `create` as standard
1. it doesn't capture exception and simply let it bubble up
2. it doesn't do `logging.info`, only `logger.debug`

---

resolved in #2